File: src/Messaging.ts
import fs from "fs";
import path from "path";

const PREFIX = "[ClaudeGates]";
const NOTIFICATION_FILE = ".pipeline-notifications";

export class Messaging
{
  public static fmt(emoji: string, text: string,): string
  {
    return `${PREFIX} ${emoji} ${text}`;
  }

  public static block(emoji: string, text: string,): void
  {
    const msg = Messaging.fmt(emoji, text,);
    process.stdout.write(JSON.stringify({ decision: "block", reason: msg, },),);
  }

  public static info(emoji: string, text: string,): void
  {
    const msg = Messaging.fmt(emoji, text,);
    process.stdout.write(JSON.stringify({ systemMessage: msg, },),);
  }

  public static notify(sessionDir: string, emoji: string, text: string,): void
  {
    const msg = Messaging.fmt(emoji, text,);
    const filePath = path.join(sessionDir, NOTIFICATION_FILE,);
    try
    {
      fs.appendFileSync(filePath, msg + "\n", "utf-8",);
    }
    catch
    {
    }
  }

  public static drainNotifications(sessionDir: string,): string | null
  {
    const filePath = path.join(sessionDir, NOTIFICATION_FILE,);
    try
    {
      if (!fs.existsSync(filePath,))
      {
        return null;
      }
      const content = fs.readFileSync(filePath, "utf-8",).trim();
      fs.unlinkSync(filePath,);
      return content || null;
    }
    catch
    {
      return null;
    }
  }

  public static log(emoji: string, text: string,): void
  {
    process.stderr.write(Messaging.fmt(emoji, text,) + "\n",);
  }

  public static readonly NOTIFICATION_FILE = NOTIFICATION_FILE;
}

File: src/claude_gates/messaging.py
import os
import sys

PREFIX = "[ClaudeGates]"
NOTIFICATION_FILE = ".pipeline-notifications"


def fmt(emoji: str, text: str) -> str:
    return f"{PREFIX} {emoji} {text}"


def block(emoji: str, text: str) -> dict:
    return {"decision": "block", "reason": fmt(emoji, text)}


def info(emoji: str, text: str) -> dict:
    return {"systemMessage": fmt(emoji, text)}


def notify(session_dir: str, emoji: str, text: str) -> None:
    msg = fmt(emoji, text)
    file_path = os.path.join(session_dir, NOTIFICATION_FILE)
    try:
        with open(file_path, "a", encoding="utf-8") as f:
            f.write(msg + "\n")
    except OSError as e:
        log("⚠️", f"notify failed: {e}")


# Note: exists+open+unlink is not truly atomic (TOCTOU). Acceptable since hooks run sequentially.
def drain_notifications(session_dir: str) -> "str | None":
    file_path = os.path.join(session_dir, NOTIFICATION_FILE)
    if not os.path.exists(file_path):
        return None
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        os.unlink(file_path)
        return content or None  # Match TS: content || null — empty string is falsy, returns None
    except OSError:
        return None


def log(emoji: str, text: str) -> None:
    sys.stderr.write(fmt(emoji, text) + "\n")
